File: samples/client/petstore/java/src/main/java/org/openapijsonschematools/schemas/validation/OneOfValidator.java
package org.openapijsonschematools.schemas.validation;

import org.openapijsonschematools.exceptions.ValidationException;

import java.util.ArrayList;
import java.util.List;

public class OneOfValidator implements KeywordValidator {
    public final List<Class<? extends JsonSchema>> oneOf;

    public OneOfValidator(List<Class<? extends JsonSchema>> oneOf) {
        this.oneOf = oneOf;
    }

    @Override
    public Object getConstraint() {
        return oneOf;
    }

    @Override
    public PathToSchemasMap validate(Class<? extends JsonSchema> cls, Object arg, ValidationMetadata validationMetadata, Object extra) {
        PathToSchemasMap pathToSchemas = new PathToSchemasMap();
        List<Class<? extends JsonSchema>> validatedOneOfClasses = new ArrayList<>();
        for(Class<? extends JsonSchema> oneOfClass: oneOf) {
            if (oneOfClass == cls) {
                /*
                optimistically assume that cls schema will pass validation
                do not invoke validate on it because that is recursive
                */
                validatedOneOfClasses.add(oneOfClass);
                continue;
            }
            try {
                PathToSchemasMap otherPathToSchemas = JsonSchema.validate(oneOfClass, arg, validationMetadata);
                validatedOneOfClasses.add(oneOfClass);
                pathToSchemas.update(otherPathToSchemas);
            } catch (ValidationException e) {
                // silence exceptions because the code needs to accumulate validatedOneOfClasses
            }
        }
        if (validatedOneOfClasses.isEmpty()) {
            throw new ValidationException("Invalid inputs given to generate an instance of "+cls+". None "+
                    "of the oneOf schemas matched the input data."
            );
        }
        if (validatedOneOfClasses.size() > 1) {
            throw new ValidationException("Invalid inputs given to generate an instance of "+cls+". Multiple "+
                    "oneOf schemas validated the data, but a max of one is allowed."
            );
        }
        return pathToSchemas;
    }
}

File: samples/client/petstore/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/OneOfValidator.kt
package org.openapijsonschematools.client.schemas.validation;

import org.openapijsonschematools.client.exceptions.ValidationException;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.ArrayList;
import java.util.List;

public class OneOfValidator implements KeywordValidator {
    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) throws ValidationException {
        var oneOf = data.schema().oneOf;
        if (oneOf == null) {
            return null;
        }
        PathToSchemasMap pathToSchemas = new PathToSchemasMap();
        List<Class<? extends JsonSchema<?>>> validatedOneOfClasses = new ArrayList<>();
        for(Class<? extends JsonSchema<?>> oneOfClass: oneOf) {
            if (oneOfClass == data.schema().getClass()) {
                /*
                optimistically assume that schema will pass validation
                do not invoke validate on it because that is recursive
                */
                validatedOneOfClasses.add(oneOfClass);
                continue;
            }
            try {
                JsonSchema<?> oneOfSchema = JsonSchemaFactory.getInstance(oneOfClass);
                PathToSchemasMap otherPathToSchemas = JsonSchema.validate(oneOfSchema, data.arg(), data.validationMetadata());
                validatedOneOfClasses.add(oneOfClass);
                pathToSchemas.update(otherPathToSchemas);
            } catch (ValidationException e) {
                // silence exceptions because the code needs to accumulate validatedOneOfClasses
            }
        }
        if (validatedOneOfClasses.isEmpty()) {
            throw new ValidationException("Invalid inputs given to generate an instance of "+data.schema().getClass()+". None "+
                    "of the oneOf schemas matched the input data."
            );
        }
        if (validatedOneOfClasses.size() > 1) {
            throw new ValidationException("Invalid inputs given to generate an instance of "+data.schema().getClass()+". Multiple "+
                    "oneOf schemas validated the data, but a max of one is allowed."
            );
        }
        return pathToSchemas;
    }
}
